cpp/src/arrow/engine/substrait/util.cc

```diff
@@ -0,0 +1,100 @@
+
+namespace arrow {
+
+namespace engine {
+
+Status SubstraitSinkConsumer::Consume(cp::ExecBatch batch) {
+  // Consume a batch of data
+  bool did_push = producer_.Push(batch);
+  if (!did_push) return Status::ExecutionError("Producer closed already");
+  return Status::OK();
+}
+
+Future<> SubstraitSinkConsumer::Finish() {
+  producer_.Push(IterationEnd<arrow::util::optional<cp::ExecBatch>>());
+  if (producer_.Close()) {
+    return Future<>::MakeFinished();
+  }
+  return Future<>::MakeFinished(
+      Status::ExecutionError("Error occurred in closing the batch producer"));
+}
+
+Status SubstraitExecutor::MakePlan() {
+  ARROW_ASSIGN_OR_RAISE(auto serialized_plan,
+                        engine::internal::SubstraitFromJSON("Plan", substrait_json_));
+  RETURN_NOT_OK(engine::internal::SubstraitToJSON("Plan", *serialized_plan));
+  std::vector<std::shared_ptr<cp::SinkNodeConsumer>> consumers;
+  std::function<std::shared_ptr<cp::SinkNodeConsumer>()> consumer_factory = [&] {
+    consumers.emplace_back(new SubstraitSinkConsumer{generator_});
+    return consumers.back();
+  };
+  ARROW_ASSIGN_OR_RAISE(declarations_,
+                        engine::DeserializePlan(*serialized_plan, consumer_factory));
+  return Status::OK();
+}
+
+arrow::PushGenerator<arrow::util::optional<cp::ExecBatch>>::Producer
+SubstraitSinkConsumer::MakeProducer(
+    AsyncGenerator<arrow::util::optional<cp::ExecBatch>>* out_gen,
+    arrow::util::BackpressureOptions backpressure) {
+  arrow::PushGenerator<arrow::util::optional<cp::ExecBatch>> push_gen(
+      std::move(backpressure));
+  auto out = push_gen.producer();
+  *out_gen = std::move(push_gen);
+  return out;
+}
+
+Result<std::shared_ptr<RecordBatchReader>> SubstraitExecutor::Execute() {
+  RETURN_NOT_OK(SubstraitExecutor::MakePlan());
+  for (const cp::Declaration& decl : declarations_) {
+    RETURN_NOT_OK(decl.AddToPlan(plan_.get()).status());
+  }
+  ARROW_RETURN_NOT_OK(plan_->Validate());
+  ARROW_RETURN_NOT_OK(plan_->StartProducing());
+  auto schema = plan_->sinks()[0]->output_schema();
+  std::shared_ptr<RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+      schema, std::move(*generator_), exec_context_.memory_pool());
+  return sink_reader;
+}
+
+Status SubstraitExecutor::Close() {
+  ARROW_RETURN_NOT_OK(plan_->finished().status());
+  return Status::OK();
+}
+
+Result<std::shared_ptr<RecordBatchReader>> SubstraitExecutor::GetRecordBatchReader(
+    std::string& substrait_json) {
+  cp::ExecContext exec_context(arrow::default_memory_pool(),
+                               ::arrow::internal::GetCpuThreadPool());
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+  ARROW_ASSIGN_OR_RAISE(auto plan, cp::ExecPlan::Make());
+  arrow::engine::SubstraitExecutor executor(substrait_json, &sink_gen, plan,
+                                            exec_context);
+  RETURN_NOT_OK(executor.MakePlan());
+  ARROW_ASSIGN_OR_RAISE(auto sink_reader, executor.Execute());
```

Comment:
Minor nit: It might be slightly simpler to have `executor.Execute()` make the call to `MakePlan` and `Close`.  That way `SubstraitExecutor` only has one public method there's less coupling between this static method and the class.

On the other hand, if you want to keep it this way, I would change the method `MakePlan` to be named `Init`.